### PR TITLE
1661 Mapping Multiple Outputs Backend

### LINKED ISSUE #1661: Mapping Table with multiple outputs on Backend
The changes needed to make on the back-end side so that the Mapping Rule is able to return multiple columns.

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Closes #1661
Implementations for Broadcast and GroupExplode

### CODE REVIEW

**Review by Zejnilovic — COMMENTED** [review]
I like the amount of tests. 
I Will still want to look into the Mapping interpreters.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala]: Now it may be a bit debattable, since they are not so large anymore

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/DynamicInterpreter.scala:238]: This feature is being implemented for `MappingRuleInterpreterGroupExplode` and `MappingruleInterpreterBroadcast`. I added #1696 in case it may be taken in the future, but it's rather difficult to do it and I would say `MappingRuleInterpreterGroupExplode` is not really experimental anymore

File: pom.xml
Comment (AdrianOlosutean) [inline pom.xml:183]: I will wait for the next spark-hats release

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:139]: I totally agree. If others agree, I will remove it

Comment (Zejnilovic) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:139]: We cannot remove it just like that. Systems depend on it. But we should move to deprecate in a proper way

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:64]: @benedeki @dk1844 Opinions?

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala
Comment (dk1844) [inline utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala:204]: for example:
```suggestion
        val defaultValue = defaultValueMap.get(mappingTable.value.outputColumns.values.head).orNull
        (param1: Any => {
          val mt = mappingTable.value.map
          mt.getOrElse(Seq(param1), defaultValue)
        })
```

This seems to be common to all 1-10 param methods.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingInterpreterSuite.scala
Comment (dk1844) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingInterpreterSuite.scala:41]: I can see that this is was in the existing code, too, but I am missing the information why it is ok to disregard this difference. Naively, this seems like a possible place to miss a schema difference. Maybe brief information on this could be present here or at the place of usage?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala:164]: I think this part be extracted to a local def to make the rest of the code more readable.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala:193]: dtto?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterGroupExplode.scala:88]: This method's code does not seem particularly readable and maintainable. Perhaps it could get split somehow to apply different LoDs. Even though all the individual parts make sense, together it's just seems to solve too many things at once.

**Review by dk1844 — DISMISSED** [review]
I didn't find any big problems, just some hint for improvement. (just read through)

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:49]: What do you mean?

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/LocalMappingTable.scala
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/LocalMappingTable.scala:85]: I was thinking the same but postponed it. I should figure it now

File: spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingInterpreterSuite.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingInterpreterSuite.scala:41]: @benedeki ?

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/LocalMappingTable.scala
Comment (dk1844) [reply utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/LocalMappingTable.scala]: Much better now.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:64]: Yup, I agree with @Zejnilovic. IMHO best place is `za.co.absa.enceladus.utils.schema.SchemaUtils`.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:49]: I guess, the question is, why is this a `val`. It can be easilly be `def`, actually considering it's used only at one place can as well be a local `val`.

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala
Comment (AdrianOlosutean) [reply utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala:170]: I created #1741 and I will remove this comment

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:140]: Redundant braces

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:149]: Redundant braces

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:157]: Misaligned braces

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:138]: I would move this `val` creation right before its use,

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:148]: To preserve the original logic here should be a check for non-empty `defaultValuesForTargets`

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:140]: This is probably the matter of taste, but wouldn't this be more understandable, what is going on in the code?
```suggestion
    val defaultValuesForTargets = rule.allOutputColumns().values.toSet.flatMap{targetAttribute => {
```

**Review by benedeki — CHANGES_REQUESTED** [review]

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/CommonMappingRuleInterpreter.scala:139]: I guess it's about the `*` expression for defaults

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala
Comment (benedeki) [inline utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala:190]: @AdrianOlosutean hey let's have a call/pair programming session over:
* `private def getMappingLambdaParamX(_, _, false)`
* `private def mapMultipleOutputs`
* `final case class LocalMappingTable`
I would like to discus some changes, imho improvements, but maybe not. And too long to describe and discuss via chat here.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala:62]: Yes, I created #1697. I will remove it

**Review by benedeki — DISMISSED** [review]

**Review by benedeki — COMMENTED** [review]
I think the `BroadcastUtils` look much better now. IMHO would be still worth to look at `LocalMappingTable`.
My approval is to unblock the eventual merge, don't take it as a done approval. But there is nothing really blocking there.
Please @dk1844 test the feature, particularly in cluster, to see the changes are streamable.

// @dk1844: @AdrianOlosutean informed me that the matter was consulted with @kevinwallimann and the current coverage is sufficient. Feel free to 'reopen' the request if necessary, @benedeki.

File: utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala
Comment (dk1844) [reply utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/BroadcastUtils.scala]: I really like what you have done in this regard, BroadcastUtils do seem much tidier now 👍

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala
Comment (dk1844) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala:50]:
```suggestion
    if (additionalColumns.getOrElse(Map()).isEmpty) {
```

Btw. additionalColumns emptyness is also check at `MappingRuleInterpreterGroupExplode:54` and at `DynamicInterpreter:239` only using `.isEmpty`, so I am wondering if this is enough or perhaps it should also be checked for `Some(Map())`, too.

**Review by dk1844 — APPROVED** [review]
LGTM.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala
Comment (dk1844) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala]: And the condition at `DynamicInterpreter:239` is non-issue? If so, why btw?

Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/conformance/interpreter/rules/mapping/MappingRuleInterpreterBroadcast.scala]: I forgot about it, I will push that too

**Review by dk1844 — DISMISSED** [review]

**Review by dk1844 — APPROVED** [review]

**Review by HuvarVer — APPROVED** [review]
Did some functionality tests with setting new options with frontend and run conformance with multiple output mapping conformance rule.

**Pull request merged**
